runtime/drivers/mysql/sql_store.go

```diff
@@ -0,0 +1,189 @@
+	"database/sql"
+	sqldriver "database/sql/driver"
+	"errors"
+	"fmt"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/mitchellh/mapstructure"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime/drivers"
+)
+
+// Query implements drivers.SQLStore
+func (c *connection) Query(ctx context.Context, props map[string]any) (drivers.RowIterator, error) {
+	srcProps, err := parseSourceProperties(props)
+	if err != nil {
+		return nil, err
+	}
+
+	var dsn string
+	if srcProps.DSN != "" { // get from src properties
+		dsn = srcProps.DSN
+	} else if url, ok := c.config["dsn"].(string); ok && url != "" { // get from driver configs
+		dsn = url
+	} else {
+		return nil, fmt.Errorf("the property 'dsn' is required for MySQL. Provide 'dsn' in the YAML properties or pass '--env connector.mysql.dsn=...' to 'rill start'")
+	}
+
+	conf, err := mysql.ParseDSN(dsn)
+	if err != nil {
+		return nil, err
+	}
+	conf.ParseTime = true // if set to false, time is scanned as an array rather than as time.Time
+
+	db, err := sql.Open("mysql", conf.FormatDSN())
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate DSN data:
+	err = db.Ping()
+	if err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	rows, err := db.QueryContext(ctx, srcProps.SQL)
+	if err != nil {
+		return nil, err
+	}
+
+	iter := &rowIterator{
+		db:   db,
+		rows: rows,
+	}
+
+	if err := iter.setSchema(); err != nil {
+		iter.Close()
+		return nil, err
+	}
+	return iter, nil
+}
+
+// QueryAsFiles implements drivers.SQLStore
+func (c *connection) QueryAsFiles(ctx context.Context, props map[string]any, opt *drivers.QueryOption, p drivers.Progress) (drivers.FileIterator, error) {
+	return nil, drivers.ErrNotImplemented
+}
+
+type rowIterator struct {
+	db   *sql.DB
+	rows *sql.Rows
+
+	schema       *runtimev1.StructType
+	row          []sqldriver.Value
+	fieldMappers []mapper
+	fieldDests   []any // Destinations are used while scanning rows
+}
+
+// Close implements drivers.RowIterator.
+func (r *rowIterator) Close() error {
+	r.rows.Close()
+	r.db.Close()
+	return nil
+}
+
+// Next implements drivers.RowIterator.
+func (r *rowIterator) Next(ctx context.Context) ([]sqldriver.Value, error) {
+	if !r.rows.Next() {
+		err := r.rows.Err()
+		if err == nil {
+			return nil, drivers.ErrIteratorDone
+		}
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, fmt.Errorf("no results found for the query")
+		}
+		return nil, err
+	}
+
+	cts, err := r.rows.ColumnTypes()
+	if err != nil {
+		return nil, err
+	}
+
+	// Scan expects destinations to be pointers
+	for i := range r.fieldDests {
+		r.fieldDests[i], err = r.fieldMappers[i].dest(cts[i].ScanType())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err := r.rows.Scan(r.fieldDests...); err != nil {
+		return nil, err
+	}
+
+	for i := range r.schema.Fields {
+		// Dereference destinations and fill the row
+		r.row[i], err = r.fieldMappers[i].value(r.fieldDests[i])
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
I think there might still be room to simplify this code. You can pass an `any` value to `Scan`. What if we just did something like:
```go
// in setSchema
r.dests = make([]any, len(r.schema.Fields))
for i := range r.dests {
  r.dests[i] = new(any)
}

// in Scan
err = r.rows.Scan(r.dests...)
if err != nil {...}

for i := r.schema.Fields {
  r.row[i], err = r.mappers[i].value(dests[i])
}
```

Reply:
This requires implementing some conversions that `Scan` has. For example, numeric types are represented as `[]uint8`. But to use `Scan`'s conversions one needs to appropriately prepare destinations - `dest` function

Reply:
Okay, that's a shame, but makes sense